grakn-engine/src/main/java/ai/grakn/engine/postprocessing/ConceptFixer.java

```diff
@@ -43,53 +51,122 @@
+        String jobId = UUID.randomUUID().toString();
         boolean notDone = true;
         int retry = 0;
+
         while (notDone) {
-            try (AbstractGraknGraph graph = (AbstractGraknGraph) EngineGraknGraphFactory.getInstance().getGraph(keyspace)) {
-                if (graph.fixDuplicateCasting(castingId)) {
-                    graph.commit((c, r) -> {}); //No caching is needed in this case
-                }
-                cache.deleteJobCasting(graph.getKeyspace(), castingId);
+            //Try to Fix the job
+            try(EngineGraknGraph graph = EngineGraknGraphFactory.getInstance().getGraph(keyspace))  {
+
+                //Perform the fix
+                BiConsumer<EngineCache, String> jobFinaliser = postProcessor.apply(graph, conceptIds);
+
+                //Check if the fix worked
+                validateFix(graph, conceptIndex, conceptIds).
+                        ifPresent(message -> {
+                            throw new RuntimeException(message);
+                        });
+
+                //Finally clear the cache
+                jobFinaliser.accept(EngineCacheImpl.getInstance(), conceptIndex);
+
+                return true; //If it can get here. All post processing has succeeded.
+            } catch (Throwable t) { //These exceptions need to become more specialised
+                LOG.warn(ErrorMessage.POSTPROCESSING_ERROR.getMessage(jobId, t.getMessage()), t);
+            }
+
+            //Fixing the job has failed after several attempts
+            if (retry > MAX_RETRY) {
                 notDone = false;
-            } catch (Exception e) {
-                LOG.warn(ErrorMessage.POSTPROCESSING_ERROR.getMessage("casting", e.getMessage()), e);
-                if (retry++ > MAX_RETRY) {
-                    LOG.error(ErrorMessage.UNABLE_TO_ANALYSE_CONCEPT.getMessage(castingId, e.getMessage()), e);
-                    notDone = false;
-                } else {
-                    performRetry(retry);
+            }
+
+            retry = performRetry(retry);
+        }
+
+        StringBuilder failingConcepts = new StringBuilder();
+        conceptIds.stream().map(id -> failingConcepts.append(id.getValue()).append(","));
+        LOG.error(ErrorMessage.UNABLE_TO_ANALYSE_CONCEPT.getMessage(failingConcepts, jobId));
+        return false;
+    }
+
+    /**
+     * Checks that post processing was done successfully by doing two things:
+     *  1. That there is only 1 valid conceptID left
+     *  2. That the concept Index does not return null
+     * @param graph A grakn graph to run the checks against.
+     * @param conceptIndex The concept index which MUST return a valid concept
+     * @param conceptIds The concpet ids which should only return 1 valid concept
+     * @return An error if one of the above rules are not satisfied.
+     */
+    private static Optional<String> validateFix(EngineGraknGraph graph, String conceptIndex, Set<ConceptId> conceptIds){
```

Comment:
So this is really checking that all the concepts were merged? Perhaps it should be `validateMerged`.